web/elm/src/Causality/Causality.elm

```diff
@@ -0,0 +1,584 @@
+                        , outgoing = IntDict.insert parentId () ctx.outgoing
+                    }
+
+        updateNode : Maybe (NodeContext NodeType ()) -> Maybe (NodeContext NodeType ())
+        updateNode nodeContext =
+            Just <|
+                case nodeContext of
+                    Just { node, incoming, outgoing } ->
+                        let
+                            oldNode =
+                                node.label
+
+                            newNodeType : NodeType
+                            newNodeType =
+                                case oldNode of
+                                    Job name builds ->
+                                        insert builds { id = b.id, name = b.name, status = b.status }
+                                            |> List.sortBy .id
+                                            |> List.reverse
+                                            |> Job name
+
+                                    _ ->
+                                        oldNode
+                        in
+                        addEdge
+                            { node =
+                                { node | label = newNodeType }
+                            , incoming = incoming
+                            , outgoing = outgoing
+                            }
+
+                    Nothing ->
+                        addEdge
+                            { node =
+                                { id = nodeId
+                                , label = Job b.jobName [ { id = b.id, name = b.name, status = b.status } ]
+                                }
+                            , incoming = IntDict.empty
+                            , outgoing = IntDict.empty
+                            }
+
+        updatedGraph =
+            Graph.update nodeId updateNode graph
+    in
+    List.foldl (\build acc -> constructResourceVersion nodeId dir build acc) updatedGraph b.resourceVersions
+
+
+constructGraph : CausalityDirection -> CausalityResourceVersion -> Graph NodeType ()
+constructGraph direction rv =
+    let
+        graph =
+            constructResourceVersion 0 direction rv Graph.empty
+    in
+    -- because the first node is constructed with fictious parentId 0, the first edge needs to be removed
+    Graph.update rv.versionId
+        (Maybe.map
+            (\ctx ->
+                case direction of
+                    Downstream ->
+                        { ctx | incoming = IntDict.remove 0 ctx.incoming }
+
+                    Upstream ->
+                        { ctx | outgoing = IntDict.remove 0 ctx.outgoing }
+            )
+        )
+        graph
+
+
+
+-- http://www.graphviz.org/doc/info/shapes.html#html. this should probably use Json.Encode.string to sanitize the output
+
+
+attributes : List ( String, String ) -> String
+attributes =
+    List.map (\( k, v ) -> k ++ "=\"" ++ v ++ "\"") >> String.join " "
```

Comment:
They value (and probably key to be safe, though we're the ones who pick the keys) need to be escaped since they are encoded in the pseudo-html, and can break things otherwise, e.g.

```elm
escape : String -> String
escape =
    String.replace "<" "&lt;"
        >> String.replace ">" "&gt;"
        >> String.replace "&" "&amp;"
        >> String.replace "\"" "&quot;"
        >> String.replace "'" "&#039;"


attributes : List ( String, String ) -> String
attributes =
    List.map (\( k, v ) -> escape k ++ "=\"" ++ escape v ++ "\"") >> String.join " "
```

Note that this no longer requires `Json.Encode.string` to avoid the value containing a quote since we escape it properly

---
Actually, maybe we want an `htmlAttributes` that does this escaping, since this escaping isn't valid for "normal" attributes - but I suppose it doesn't really matter too much

Reply:
> Actually, maybe we want an htmlAttributes that does this escaping

Not quite sure I know what this means. Either way, I think we'll need to keep this function because even the cell text also needs to be escaped (i.e. `<TD>&lt;special&amp;name&gt;</TD>`)

Reply:
Right, I guess I just meant the way that characters get escaped when they are just regular attributes (as opposed to "HTML" attributes) is different. For instance, `"` should become `&quot;` in HTML, but `\"` in regular attributes.

That said, we don't currently really *need* to escape anything in the regular attributes, though, so that's why I say it doesn't matter too much

Reply:
But in any event, c85181d215df0676700d7ce89af89026d722a5c6 isn't sufficient for making things not break - e.g. consider https://ci.concourse-ci.org/teams/main/pipelines/aidan/resources/mock/causality/942884455/downstream

<img width="774" alt="Screen Shot 2021-06-04 at 1 34 44 PM" src="https://user-images.githubusercontent.com/26583442/120841552-a2cbb900-c539-11eb-8ded-12d5512ff514.png">